README.md


# MarkowitzR

[![Build Status](https://github.com/shabbychef/MarkowitzR/workflows/R-CMD-check/badge.svg)](https://github.com/shabbychef/MarkowitzR/actions)
[![codecov.io](http://codecov.io/github/shabbychef/MarkowitzR/coverage.svg?branch=master)](https://app.codecov.io/github/shabbychef/MarkowitzR?branch=master)
[![CRAN](http://www.r-pkg.org/badges/version/MarkowitzR)](https://cran.r-project.org/package=MarkowitzR)
[![Downloads](http://cranlogs.r-pkg.org/badges/MarkowitzR?color=green)](https://www.r-pkg.org:443/pkg/MarkowitzR)
[![Total](http://cranlogs.r-pkg.org/badges/grand-total/MarkowitzR?color=green)](https://www.r-pkg.org:443/pkg/MarkowitzR)



A number of utilities for dealing with the Markowitz portfolio.

-- Steven E. Pav, shabbychef@gmail.com

## Installation

This package may be installed from CRAN; the latest version may be
found on [github](https://github.com/shabbychef/MarkowitzR "MarkowitzR")
via devtools:


```r
if (require(devtools)) {
    # latest greatest
    install_github(repo = "MarkowitzR", username = "shabbychef", 
        ref = "master")
}
```

# Basic Usage

## Inference on the Markowitz Portfolio

The (negative) Markowitz portfolio appears in the inverse of
the uncentered second moment matrix of the 'augmented' vector
of returns. Via the Central Limit Theorem and the delta method
the asymptotic distribution of the Markowitz portfolio can
be found. From this, Wald statistics on the individual portfolio
weights can be computed. 

### Fake Data

First for unconditional returns:


```r
set.seed(1001)
X <- matrix(rnorm(1000 * 3), ncol = 3)
ism <- mp_vcov(X, fit.intercept = TRUE)
walds <- ism$W/sqrt(diag(ism$What))
print(t(walds))
```

```
##             X1    X2   X3
## Intercept 0.83 -0.15 -1.8
```

Now for conditional expectation:


```r
# generate data with given W, Sigma
Xgen <- function(W, Sigma, Feat) {
    Btrue <- Sigma %*% W
    Xmean <- Feat %*% t(Btrue)
    Shalf <- chol(Sigma)
    X <- Xmean + matrix(rnorm(prod(dim(Xmean))), ncol = dim(Xmean)[2]) %*% 
        Shalf
}

n.feat <- 3
n.ret <- 5
n.obs <- 2000
set.seed(101)
Feat <- matrix(rnorm(n.obs * n.feat), ncol = n.feat)
Wtrue <- 5 * matrix(rnorm(n.feat * n.ret), ncol = n.feat)
Sigma <- cov(matrix(rnorm(100 * n.ret), ncol = n.ret))
Sigma <- Sigma + diag(seq(from = 1, to = 3, length.out = n.ret))
X <- Xgen(Wtrue, Sigma, Feat)
ism <- mp_vcov(X, feat = Feat, fit.intercept = TRUE)

walds <- ism$W/sqrt(diag(ism$What))
print(t(walds))
```

```
##               X1    X2     X3     X4    X5
## Intercept  -0.61   0.4   0.11   0.61   0.2
## Feat1       0.30 -20.7  14.56 -10.70 -24.6
## Feat2     -14.82  -4.0  12.01  29.75  -1.8
## Feat3      16.62  -5.3 -30.41   0.83   3.5
```

```r
# results are not much changed when using robust
# s.e.
library(sandwich)

ism.rse <- mp_vcov(X, feat = Feat, vcov.func = sandwich::vcovHAC, 
    fit.intercept = TRUE)
walds.rse <- ism.rse$W/sqrt(diag(ism.rse$What))
print(t(walds.rse))
```

```
##               X1    X2     X3     X4     X5
## Intercept  -0.61   0.4   0.11   0.60   0.19
## Feat1       0.29 -20.4  14.43 -10.66 -24.45
## Feat2     -14.49  -3.9  11.88  29.60  -1.80
## Feat3      16.37  -5.3 -29.85   0.83   3.48
```

```r
# errors should be asymptotically normal with the
# given covariance.
n.feat <- 5
n.ret <- 15
n.obs <- 3000
set.seed(101)

Feat <- matrix(rnorm(n.obs * n.feat), ncol = n.feat)
Wtrue <- 5 * matrix(rnorm(n.feat * n.ret), ncol = n.feat)
Sigma <- cov(matrix(rnorm(100 * n.ret), ncol = n.ret))
Sigma <- Sigma + diag(seq(from = 1, to = 3, length.out = n.ret))
X <- Xgen(Wtrue, Sigma, Feat)
ism <- mp_vcov(X, feat = Feat, fit.intercept = TRUE)

Wcomp <- cbind(0, Wtrue)
errs <- ism$W - Wcomp
dim(errs) <- c(length(errs), 1)
Zerr <- solve(t(chol(ism$What)), errs)
print(summary(Zerr))
```

```
##        V1       
##  Min.   :-2.05  
##  1st Qu.:-0.55  
##  Median : 0.14  
##  Mean   : 0.20  
##  3rd Qu.: 1.00  
##  Max.   : 2.49
```

```r
library(ggplot2)
ph <- ggplot(data.frame(Ze = Zerr), aes(sample = Ze)) + 
    stat_qq() + geom_abline(slope = 1, intercept = 0, 
    colour = "red")
print(ph)
```

<img src="man/figures/marko-ism-1.png" title="plot of chunk marko-ism" alt="plot of chunk marko-ism" width="600px" height="500px" />

```r
# qqnorm(Zerr) qqline(Zerr,col=2)
```

### Fama French data

Now load the Fama French 3 factor portfolios.


```r
if (!require(aqfb.data, quietly = TRUE) && require(devtools)) {
    # get the 10 industry data
    devtools::install_github("shabbychef/aqfb_data")
}
library(aqfb.data)
# fama
data(mff4)

# will not matter, but convert pcts:
ff.data <- 0.01 * mff4

# risk free rate:
rfr <- ff.data[, "RF"]

# subtract risk free from Mkt, HML and SMB:
ff.ret <- ff.data[, c("Mkt", "HML", "SMB")] - rep(rfr, 
    2)
```

Now analyze the Markowitz portfolio on them.


```r
ism <- mp_vcov(ff.ret, fit.intercept = TRUE)
walds <- ism$W/sqrt(diag(ism$What))
print(t(walds))
```

```
##           Mkt  HML SMB
## Intercept   4 0.32  -2
```

```r
# now consider the hedging constraint: no
# covariance with the market:
Gmat <- matrix(c(1, 0, 0), nrow = 1)
ism <- mp_vcov(ff.ret, fit.intercept = TRUE, Gmat = Gmat)
walds <- ism$W/sqrt(diag(ism$What))
print(t(walds))
```

```
##           Mkt  HML SMB
## Intercept 1.5 0.32  -2
```



ChangeLog

2023-08-20 Steven E. Pav <shabbychef@gmail.com>
* version 1.0.3 on CRAN
* fix package documentation

2020-01-07 Steven E. Pav <shabbychef@gmail.com>
* version 1.0.2 on CRAN
* maintenance release to fix atlas errors?

2018-05-25 Steven E. Pav <shabbychef@gmail.com>
* move figures for README on CRAN

2015-01-26 Steven E. Pav <shabbychef@gmail.com>
* conform to CRAN rules

2014-02-14 Steven E. Pav <shabbychef@gmail.com>
* first release

2014-01-04 Steven E. Pav <shabbychef@gmail.com>
* start work



R/MarkowitzR.r
# Copyright 2014-2020 Steven E. Pav. All Rights Reserved.
# Author: Steven E. Pav
#
# This file is part of MarkowitzR.
#
# MarkowitzR is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# MarkowitzR is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with MarkowitzR.  If not, see <http://www.gnu.org/licenses/>.

# Created: 2014.01.31
# Copyright: Steven E. Pav, 2014
# Author: Steven E. Pav
# Comments: Steven E. Pav

#' Inference on the Markowitz portfolio.
#' 
#' @section Markowitz Portfolio:
#'
#' Suppose \eqn{x} is a \eqn{p}-vector of returns of some assets with expected
#' value \eqn{\mu}{mu} and covariance \eqn{\Sigma}{Sigma}. The 
#' \emph{Markowitz Portfolio} is the portfolio 
#' \eqn{w = \Sigma^{-1}\mu}{w = Sigma^-1 mu}. Scale multiples of this portfolio
#' solve various portfolio optimization problems, among them
#' \deqn{\mathrm{argmax}_{w: w^{\top}\Sigma w \le R^2} \frac{\mu^{\top} w -
#' r_0}{\sqrt{w^{\top}\Sigma w}}}{argmax{ (mu'w - r0) / sqrt(w'Sigma w) :
#' w'Sigma w <= R^2}}
#'
#' This packages supports various statistical tests around the elements of 
#' the Markowitz Portfolio, and its Sharpe ratio, including the possibility of
#' hedging, and scalar conditional heteroskedasticity and conditional
#' expectation.
#'
#' @section Legal Mumbo Jumbo:
#'
#' MarkowitzR is distributed in the hope that it will be useful,
#' but WITHOUT ANY WARRANTY; without even the implied warranty of
#' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#' GNU Lesser General Public License for more details.
#'
#' @template etc
#' @template ref-SEP13
#' @template ref-MBJ99
#' @template ref-Markowitz52
#'
#' @references
#'
#' Bodnar, Taras and Okhrin, Yarema. "On the Product of Inverse Wishart
#' and Normal Distributions with Applications to Discriminant Analysis 
#' and Portfolio Theory." Scandinavian Journal of Statistics 38, no. 2 (2011):
#' 311--331. \doi{10.1111/j.1467-9469.2011.00729.x}
#'
#' Brandt, Michael W. "Portfolio Choice Problems." Handbook of Financial
#' Econometrics 1 (2009): 269--336. 
#' \url{https://scholars.duke.edu/publication/964964}
#'
#' @import matrixcalc gtools
#' @importFrom stats lm
#' @importFrom stats na.omit
#' @importFrom stats vcov
#'
#' @name MarkowitzR
#' @rdname MarkowitzR
#' @docType package
#' @title statistics concerning the Markowitz portfolio
#' @keywords package
#' @aliases MarkowitzR-package
#' 
#' @note 
#' This package is maintained as a hobby. 
#'
"_PACKAGE"

#' @title News for package 'MarkowitzR':
#'
#' @description 
#'
#' News for package \sQuote{MarkowitzR}
#'
#' \newcommand{\CRANpkg}{\href{https://cran.r-project.org/package=#1}{\pkg{#1}}}
#' \newcommand{\MarkowitzR}{\CRANpkg{MarkowitzR}}
#'
#' @section Changes in \MarkowitzR{} Version 1.0.3 (2023-08-20) :
#' \itemize{
#' \item fix package documentation.
#' }
#'
#'
#' @section Changes in \MarkowitzR{} Version 1.0.2 (2020-01-07) :
#' \itemize{
#' \item emergency fix for broken tests under atlas on CRAN.
#' }
#'
#' @section Changes in \MarkowitzR{} Version 1.0.1 (2018-05-25) :
#' \itemize{
#' \item move figures around for README on CRAN.
#' }
#'
#' @section Changes in \MarkowitzR{} Version 0.9900 (2016-09-15) :
#' \itemize{
#' \item yet again, conform to CRAN rules.
#' }
#'
#' @section Changes in \MarkowitzR{} Version 0.1502 (2015-01-26) :
#' \itemize{
#' \item conform to CRAN rules.
#' }
#'
#' @section Changes in \MarkowitzR{} Version 0.1403 (2014-06-01) :
#' \itemize{
#' \item fix bug preventing multi-row hedging or constraint matrices.
#' }
#'
#' @section \MarkowitzR{} Initial Version 0.1402 (2014-02-14) :
#' \itemize{
#' \item first CRAN release.
#' }
#'
#' @name MarkowitzR-NEWS
#' @rdname NEWS
NULL

#for vim modeline: (do not edit)
# vim:fdm=marker:fmr=FOLDUP,UNFOLD:cms=#%s:syn=r:ft=r
